File: src/main/java/mnm/mods/tabbychat/client/gui/component/VerticalLayout.java
package mnm.mods.tabbychat.client.gui.component;

import java.util.List;

import com.google.common.collect.Lists;

import mnm.mods.tabbychat.util.Dim;
import mnm.mods.tabbychat.util.ILocation;
import mnm.mods.tabbychat.util.Location;

/**
 * Displays components top to bottom. Like {@link FlowLayout}, but vertical.
 * 
 * @author Matthew
 */
public class VerticalLayout implements ILayout {

    private List<GuiComponent> list = Lists.newArrayList();

    @Override
    public void addComponent(GuiComponent comp, Object constraints) {
        if (constraints != null) {
            if (constraints instanceof Integer) {
                list.add((Integer) constraints, comp);
            } else {
                throw new IllegalArgumentException("Illegal constraint of type: "
                        + constraints.getClass().getName() + ". Only int accepted.");
            }
        } else {
            list.add(comp);
        }
    }

    @Override
    public void removeComponent(GuiComponent comp) {
        list.remove(comp);
    }

    @Override
    public void layoutComponents(GuiPanel parent) {
        ILocation par = parent.getLocation();
        int y = par.getYPos();
        for (GuiComponent comp : list) {
            Location loc = comp.getLocation().copy();
            loc.setXPos(par.getXPos());
            loc.setYPos(y);
            comp.setLocation(loc);
            y += comp.getMinimumSize().height;
        }
    }

    @Override
    public Dim getLayoutSize() {
        int width = 0;
        int height = 0;
        for (GuiComponent comp : list) {
            width = Math.max(width, comp.getLocation().getWidth());
            height += comp.getMinimumSize().height;
        }
        return new Dim(width, height);
    }
}

File: src/main/kotlin/mnm/mods/tabbychat/client/gui/component/layout/VerticalLayout.kt
package mnm.mods.tabbychat.client.gui.component.layout

import mnm.mods.tabbychat.client.gui.component.GuiComponent
import mnm.mods.tabbychat.client.gui.component.GuiPanel
import mnm.mods.tabbychat.util.Dim
import kotlin.math.max

/**
 * Displays components top to bottom. Like [FlowLayout], but vertical.
 *
 * @author Matthew
 */
class VerticalLayout : ILayout {

    private val list = ArrayList<GuiComponent>()

    override val layoutSize: Dim
        get() {
            var width = 0
            var height = 0
            for (comp in list) {
                width = max(width, comp.location.width)
                height += comp.minimumSize.height
            }
            return Dim(width, height)
        }

    override fun addComponent(comp: GuiComponent, constraints: Any?) {
        if (constraints != null) {
            if (constraints is Int) {
                list.add((constraints as Int?)!!, comp)
            } else {
                throw IllegalArgumentException("Illegal constraint of type: ${constraints.javaClass.name}. Only int accepted.")
            }
        } else {
            list.add(comp)
        }
    }

    override fun removeComponent(comp: GuiComponent) {
        list.remove(comp)
    }

    override fun layoutComponents(parent: GuiPanel) {
        val par = parent.location
        var y = par.yPos
        for (comp in list) {
            val loc = comp.location.copy()
            loc.xPos = par.xPos
            loc.yPos = y
            comp.location = loc
            y += comp.minimumSize.height
        }
    }
}
